src/commands/dev/lib/dev-builder.ts

```diff
@@ -0,0 +1,296 @@
+  if (!buildConfig || !buildEntry) {
+    throw new Error("Asset has not been built yet, can't rebuild");
+  }
+  const { cwd } = devServer;
+  const entrypoint = relative(cwd, buildEntry.fsPath);
+
+  const cacheDir = await cacheDirPromise;
+  const { dev, ino } = await stat(entrypoint);
+  const workPath = join(cacheDir, 'workPaths', String(dev + ino));
+  await mkdirp(workPath);
+
+  devServer.logDebug(`Building ${buildEntry.fsPath} (workPath = ${workPath})`);
+  const { builder } = buildConfig;
+  if (!builder) {
+    throw new Error('No builder');
+  }
+  const builderConfig = builder.config || {};
+  const files = await collectProjectFiles('**', cwd);
+  const config = buildConfig.config || {};
+  const output = await builder.build({
+    files,
+    entrypoint,
+    workPath,
+    config,
+    isDev: true
+  });
+
+  // enforce the lambda zip size soft watermark
+  const { maxLambdaSize = '5mb' } = { ...builderConfig, ...config };
+  let maxLambdaBytes: number;
+  if (typeof maxLambdaSize === 'string') {
+    maxLambdaBytes = bytes(maxLambdaSize);
+  } else {
+    maxLambdaBytes = maxLambdaSize;
+  }
+
+  for (const asset of Object.values(output)) {
+    if (asset.type === 'Lambda') {
+      const size = asset.zipBuffer.length;
+      if (size > maxLambdaBytes) {
+        throw new Error(
+          `The lambda function size (${bytes(
+            size
+          ).toLowerCase()}) exceeds the configured limit (${bytes(
+            maxLambdaBytes
+          ).toLowerCase()}). You may increase this by supplying \`maxLambdaSize\` to the build \`config\``
+        );
+      }
+    }
+
+    asset.buildConfig = buildConfig;
+    asset.buildEntry = buildEntry;
+  }
+
+  await Promise.all(
+    Object.entries(output).map(async e => {
+      const path: string = e[0];
+      const asset: BuilderOutput = e[1];
+      if (asset.type !== 'Lambda') return;
+
+      // Tear down the previous `fun` Lambda instance for this asset
+      const oldAsset = devServer.assets[path];
+      if (oldAsset && oldAsset.type === 'Lambda' && oldAsset.fn) {
+        await oldAsset.fn.destroy();
+      }
+
+      asset.fn = await createFunction({
+        Code: { ZipFile: asset.zipBuffer },
+        Handler: asset.handler,
+        Runtime: asset.runtime,
+        Environment: {
+          Variables: {
+            // TODO: resolve secret env vars
+            ...nowJson.env,
+            ...asset.environment,
+            NOW_REGION: 'dev1'
+          }
+        }
+      });
+    })
+  );
+
+  // TODO: remove previous assets, i.e. if the rebuild has different outputs that
+  // don't include some of the previous ones. For example, deleting a page in
+  // a Next.js build
+  Object.assign(devServer.assets, output);
+}
+
+async function executeBuilds(
+  nowJson: NowConfig,
+  devServer: DevServer
+): Promise<void> {
+  const { cwd } = devServer;
+  const files = await collectProjectFiles('**', cwd);
+
+  if (!nowJson.builds) {
+    return;
+  }
```

Comment:
should we move this to the function start?

Reply:
Ya good call